testcontainers/src/core/host_port_mapping.rs

```diff
@@ -0,0 +1,169 @@
+use std::collections::BTreeMap;
+
+use crate::core::{client::Client, containers::request::Host, error::TestcontainersError};
+
+/// Checks if we're running on Docker Desktop (macOS or Windows).
+/// Docker Desktop provides `host.docker.internal` by default.
+fn is_docker_desktop() -> bool {
+    let is_desktop = cfg!(any(target_os = "macos", target_os = "windows"));
+    if is_desktop {
+        log::trace!("Platform: Docker Desktop, host.docker.internal available");
+    } else {
+        log::trace!("Platform: not Docker Desktop");
+    }
+    is_desktop
+}
+
+/// Checks if a Docker version string is at least the specified major.minor version
+fn is_docker_version_at_least(version_str: &str, min_major: u32, min_minor: u32) -> bool {
+    // Docker version strings can be like "20.10.17", "24.0.0", etc.
+    let parts: Vec<&str> = version_str.split('.').collect();
+    if parts.len() < 2 {
+        return false;
+    }
+
+    let major: u32 = parts[0].parse().unwrap_or(0);
+    let minor: u32 = parts[1].parse().unwrap_or(0);
+
+    major > min_major || (major == min_major && minor >= min_minor)
+}
+
+/// Checks if the current Docker installation supports the `host-gateway` feature.
+/// This is available in Docker 20.10+ on Linux.
+async fn supports_host_gateway(client: &Client) -> bool {
+    if !cfg!(target_os = "linux") {
+        log::trace!("Platform: not Linux, host-gateway unsupported");
+        return false;
+    }
+
+    log::trace!("Platform: Linux, checking Docker version for host-gateway");
+
+    // Check Docker version to see if it supports host-gateway
+    match client.docker_version().await {
+        Ok(Some(version_str)) => {
+            let supported = is_docker_version_at_least(&version_str, 20, 10);
+            if !supported {
+                log::warn!("Docker version {} detected on Linux - host-gateway may not be supported (requires 20.10+)", version_str);
+            }
+            supported
+        }
+        Ok(None) => {
+            // No version info available, assume it's supported but warn
+            log::warn!(
+                "Docker version not available - assuming host-gateway is supported on Linux"
+            );
+            true
+        }
+        Err(err) => {
+            // If we can't get version info, assume it's supported but warn
+            log::warn!(
+                "Failed to detect Docker version ({}), assuming host-gateway is supported on Linux",
+                err
+            );
+            true
+        }
+    }
```

Comment:
This looks like good candidate to be a method of client as main argument is `&Client`

But in order to split the logic and decouple from the `Client` I'd suggest to define some trait and impl it for the `Client` (e.g `impl HostMapping for Client { ...}`)